File: tools/xmi2slb.py
#!/usr/bin/env python3
"""xmi2slb — build a Mac-format MUSIC.SLB from the DOS release's XMI music.

The DOS release ships FRUA's soundtrack as XMIDI files in four DRIVER
arrangements (the file prefix is the sound driver, not the tune):

    ADDQ?.XMI  AdLib          (multi-channel + GM drums on ch 9)
    RODQ?.XMI  Roland MT-32   (note-identical to ADDQ)
    TYDQ?.XMI  Tandy 3-voice  (three monophonic channels 1..3)
    PCDQ?.XMI  PC speaker     (one channel)

Each driver carries the same THREE songs Q1/Q2/Q3. The Mac release's
MUSIC.SLB holds EIGHT songs for its 4-tone synth; melody-trigram
correlation against the decoded Mac patterns places the DOS songs at
Mac slots (data measured 2026-07-21, TYDQ vs MUSIC.SLB):

    Q1 -> slot 0   the title / Training-Hall theme (129/201 trigrams)
    Q2 -> slot 1   the short jingle                 (20/27)
    Q3 -> slot 7   the big theme                    (69/107)

Slots 2..6 are Mac-only compositions with no DOS source; they are
written as EMPTY songs (a lone 0xFF terminator) so jt985's range check
passes and a design event requesting them plays silence instead of
popping "Song out of range".

The Tandy arrangement is the conversion source of choice: SSI's own
musicians already reduced each tune to three monophonic square-wave
voices — exactly the shape the Mac sequencer (jt974) wants.

## The .slb bank format (from jt975/l11a2/jt974, verified against the
## Mac MUSIC.SLB)

    bank:   'SLBR' u32_file_size u8_song_count u8_pad
            (count+1) x u16 song offsets (relative to the end of the
            offset table; entry [count] = total data size)
            song data, back to back
    song:   u16 period  u8 level  u8 voice_count  5 x u16 voice offsets
            (relative to song+14), then the voice pattern streams
    voice:  (value, duration) byte PAIRS.
            value 128 = rest, < 128 = MIDI note number, 255 = end;
            value > 128 or duration bit 7 set = a command pair
            (129 re-strike, 132 set level, others skipped).
            duration bits 0-2: note value, whole..1/128 (26880 >> n)
                     bit  3:   dotted (x1.5)
                     bits 4-5: tuplet t in {2,4,6}: (d/(t+1))*t
                     bit  6:   tie into the next pair
            Every Mac voice opens with the preamble 87 00 83 00 81 LL
            82 1a (135/131/129/130 command pairs; only 129 acts on the
            4-tone synth — the others target richer drivers) and ends
            with FF FF. Preamble levels are ff/47/4e by voice index.

## Timing (verified against the matched Mac/DOS song pairs)

XMIDI runs 60 ticks per quarter note; the Mac pattern grid is 26880
ticks per whole note (6720/quarter), so ONE XMI TICK = 112 MAC TICKS
(TYDQ2's opening 72-tick note is the Mac's 8064 exactly). Tempo lives
in the song-header period: reload = period*105/7200 (l0fc4), and the
matched pairs pin

    period = 15360 * 500000 / tempo_us

(500000 us/quarter -> 15360 = the Mac's own value; TYDQ3's 560747 ->
13696 = the Mac's song-7 period to the digit).

Durations are quantized onto the l0ff2 grid by aiming each event at
its ABSOLUTE ideal tick (greedy largest-fit + ties), so rounding never
accumulates across a voice and the three voices stay in step.
"""
import os
import struct
import sys

XMI_TICK = 112                  # Mac sequencer ticks per XMI tick
PERIOD_REF = 15360              # song-header period at 500000 us/quarter
TEMPO_REF = 500000
PREAMBLE_LEVELS = (0xFF, 0x47, 0x4E, 0x47, 0x47)
MAX_VOICES = 3                  # the Mac songs are 3-voice; Tandy is 3-channel


def l0ff2(b):
    """The engine's duration decode (CODE 5 + 0x0ff2), bit-faithful."""
    d = 26880 >> (b & 7)
    if b & 8:
        d = (d >> 1) * 3
    t = (b & 48) >> 3
    if b & 48:
        d = (d // (t + 1)) * t
    return d & 0xffff


def duration_grid():
    """All distinct (ticks, code) the duration byte can express, longest
    first. Codes use bits 0..5 only (bit 6 = tie, bit 7 = command)."""
    seen = {}
    for code in range(0x40):
        d = l0ff2(code)
        if d > 0 and d not in seen:
            seen[d] = code
    return sorted(seen.items(), key=lambda kv: -kv[0])


GRID = duration_grid()
GRID_MIN = GRID[-1][0]


def quantize(target):
    """Decompose `target` Mac ticks into grid codes (greedy largest-fit).
    Returns (codes, emitted_ticks); emitted is within GRID_MIN/2 of the
    target (callers aim at absolute positions, so error never drifts)."""
    codes = []
    left = target
    while left >= GRID_MIN:
        for d, c in GRID:
            if d <= left:
                codes.append(c)
                left -= d
                break
    if left > GRID_MIN // 2:
        codes.append(GRID[-1][1])
        left -= GRID_MIN
    return codes, target - left


def parse_xmi(data):
    """Parse one XMIDI file -> (notes, tempo_us).
    notes = [(tick, channel, note, velocity, duration_ticks)].
    XMIDI deltas are bare 0x00..0x7F accumulator bytes (no VLQ) and
    Note On carries a VLQ DURATION after the velocity (no Note Offs)."""
    i = data.find(b"EVNT")
    if i < 0:
        raise ValueError("no EVNT chunk")
    ln = struct.unpack(">I", data[i + 4:i + 8])[0]
    ev = data[i + 8:i + 8 + ln]
    t = 0
    p = 0
    tempo = TEMPO_REF
    notes = []
    while p < len(ev):
        b = ev[p]
        if b < 0x80:                        # interval byte
            t += b
            p += 1
            continue
        if b == 0xFF:                       # meta
            meta = ev[p + 1]
            ln2, q = 0, p + 2
            while True:
                c = ev[q]
                ln2 = (ln2 << 7) | (c & 0x7F)
                q += 1
                if not (c & 0x80):
                    break
            if meta == 0x51:
                tempo = (ev[q] << 16) | (ev[q + 1] << 8) | ev[q + 2]
            if meta == 0x2F:
                break
            p = q + ln2
            continue
        st, ch = b & 0xF0, b & 0x0F
        if st == 0x90:                      # note on + VLQ duration
            note, vel = ev[p + 1], ev[p + 2]
            q = p + 3
            dur = 0
            while True:
                c = ev[q]
                dur = (dur << 7) | (c & 0x7F)
                q += 1
                if not (c & 0x80):
                    break
            notes.append((t, ch, note, vel, dur))
            p = q
        elif st in (0x80, 0xA0, 0xB0, 0xE0):
            p += 3
        elif st in (0xC0, 0xD0):
            p += 2
        else:
            p += 1
    return notes, tempo


def monophonize(events):
    """[(start, note, dur)] sorted -> truncate overlaps (last-note wins)."""
    out = []
    for start, note, dur in sorted(events):
        if out and out[-1][0] + out[-1][2] > start:
            prev = out[-1]
            out[-1] = (prev[0], prev[1], max(0, start - prev[0]))
        out.append((start, note, dur))
    return [e for e in out if e[2] > 0]


def voice_pattern(events, level):
    """One channel's [(xmi_start, note, xmi_dur)] -> a Mac pattern stream.
    Events are aimed at their absolute ideal Mac tick so quantization
    error never accumulates."""
    pat = bytearray((0x87, 0x00, 0x83, 0x00, 0x81, level, 0x82, 0x1A))
    pos = 0                                     # emitted Mac ticks
    for start, note, dur in events:
        ideal = start * XMI_TICK
        if ideal > pos:                         # rest-fill the gap
            codes, emitted = quantize(ideal - pos)
            for c in codes:
                pat += bytes((128, c))
            pos += emitted
        codes, emitted = quantize(dur * XMI_TICK)
        for k, c in enumerate(codes):
            tie = 0x40 if k + 1 < len(codes) else 0
            pat += bytes((note, c | tie))
        pos += emitted
    pat += b"\xff\xff"
    return bytes(pat)


def song_from_xmi(data):
    """One XMI file -> a complete .slb song record."""
    notes, tempo = parse_xmi(data)
    by_ch = {}
    for t, ch, note, vel, dur in notes:
        if ch == 9:                             # GM percussion — no pitch
            continue
        by_ch.setdefault(ch, []).append((t, note, dur))
    # busiest channels first, up to the Mac voice budget
    chans = sorted(by_ch, key=lambda c: -len(by_ch[c]))[:MAX_VOICES]
    chans.sort()                                # keep the source ordering
    voices = [voice_pattern(monophonize(by_ch[c]), PREAMBLE_LEVELS[i])
              for i, c in enumerate(chans)]
    period = int(round(PERIOD_REF * TEMPO_REF / tempo))
    hdr = struct.pack(">HBB", period, 127, len(voices))
    voffs = []
    off = 0
    for v in voices:
        voffs.append(off)
        off += len(v)
    voffs += [0] * (5 - len(voffs))
    return hdr + struct.pack(">5H", *voffs) + b"".join(voices)


def empty_song():
    """A valid song that ends on its first sequencer tick (silence)."""
    return struct.pack(">HBB5H", PERIOD_REF, 127, 1, 0, 0, 0, 0, 0) + b"\xff\xff"


# Mac slot -> DOS song number (None = Mac-only composition, left empty)
SLOT_MAP = (1, 2, None, None, None, None, None, 3)


def build_bank(xmi_by_song):
    """{1: bytes, 2: bytes, 3: bytes} (XMI file contents) -> .slb bytes."""
    songs = []
    for slot, q in enumerate(SLOT_MAP):
        songs.append(song_from_xmi(xmi_by_song[q]) if q is not None
                     else empty_song())
    offs = []
    off = 0
    for s in songs:
        offs.append(off)
        off += len(s)
    offs.append(off)
    body = struct.pack(">%dH" % len(offs), *offs) + b"".join(songs)
    total = 4 + 4 + 1 + 1 + len(body)
    return b"SLBR" + struct.pack(">IBB", total, len(songs), 0) + body


def find_xmi_set(dos_dir):
    """Locate one full Q1/Q2/Q3 set under `dos_dir` (searched recursively,
    case-insensitive). Tandy (TY) preferred — already 3 monophonic voices;
    then PC (1 voice), then RO/AD (reduced to the 3 busiest channels)."""
    found = {}
    for r, _d, files in os.walk(dos_dir):
        for f in files:
            u = f.upper()
            if u.endswith(".XMI") and len(u) == 9 and u[2:4] == "DQ":
                found[(u[:2], int(u[4]))] = os.path.join(r, f)
    for pref in ("TY", "PC", "RO", "AD"):
        if all((pref, q) in found for q in (1, 2, 3)):
            return {q: found[(pref, q)] for q in (1, 2, 3)}, pref
    return None, None


def main(argv):
    if len(argv) != 2:
        print(__doc__)
        print("usage: xmi2slb.py <dos_dir> <out.slb>")
        return 2
    paths, pref = find_xmi_set(argv[0])
    if paths is None:
        print("xmi2slb: no complete ??DQ1..3.XMI set under %s" % argv[0])
        return 1
    bank = build_bank({q: open(p, "rb").read() for q, p in paths.items()})
    with open(argv[1], "wb") as f:
        f.write(bank)
    print("xmi2slb: %s (%d bytes) from the %s arrangement (%s)"
          % (argv[1], len(bank), pref,
             ", ".join(os.path.basename(paths[q]) for q in (1, 2, 3))))
    return 0


if __name__ == "__main__":
    sys.exit(main(sys.argv[1:]))

File: src/convert/xmi2slb.c
/* xmi2slb — DOS XMIDI -> Mac MUSIC.SLB, the C port of tools/xmi2slb.py.
 *
 * Every rule below has a named twin in the Python, and the ORDER of ties is
 * part of the contract: the reference sorts (start, note, dur) tuples, picks
 * the three busiest channels stable on FIRST APPEARANCE, and rounds the
 * period half-to-even. Each of those is reproduced explicitly rather than
 * approximated, because the test is byte identity.
 *
 * Format recap (the Python's docstring has the full derivation):
 *   bank:  'SLBR' u32 size  u8 count  u8 pad  (count+1) x u16 offsets  songs
 *   song:  u16 period  u8 level(127)  u8 nvoices  5 x u16 voice offsets  voices
 *   voice: preamble 87 00 83 00 81 LL 82 1a, then (value, duration) pairs,
 *          value 128 = rest, <128 = MIDI note, FF FF = end; duration bit 6 =
 *          tie into the next pair.
 *   ONE XMI TICK = 112 Mac ticks; period = round(15360 * 500000 / tempo_us).
 */
#include "xmi2slb.h"

#define XMI_TICK     112L
#define PERIOD_REF   15360UL
#define TEMPO_REF    500000UL
#define MAX_VOICES   3
#define NSLOTS       8

static const unsigned char PREAMBLE_LEVELS[5] = { 0xFF, 0x47, 0x4E, 0x47, 0x47 };
/* Mac slot -> DOS song (1..3), 0 = Mac-only composition, left empty */
static const unsigned char SLOT_MAP[NSLOTS] = { 1, 2, 0, 0, 0, 0, 0, 3 };

/* ---- the duration grid (l0ff2, bit-faithful) ---------------------------- */

static unsigned l0ff2(unsigned b)
{
	unsigned d = 26880u >> (b & 7);
	unsigned t;
	if (b & 8)
		d = (d >> 1) * 3;
	t = (b & 48) >> 3;
	if (b & 48)
		d = (d / (t + 1)) * t;
	return d & 0xffffu;
}

static unsigned g_grid_d[0x40];
static unsigned char g_grid_c[0x40];
static int g_grid_n;
static unsigned g_grid_min;

static void grid_init(void)
{
	unsigned code;
	int i, j;
	if (g_grid_n)
		return;
	/* first code for each distinct duration, then sort by duration desc.
	 * The Python keeps insertion order on ties, but durations are DISTINCT
	 * by construction here, so a plain insertion sort is order-exact. */
	for (code = 0; code < 0x40; code++) {
		unsigned d = l0ff2(code);
		int seen = 0;
		if (d == 0)
			continue;
		for (i = 0; i < g_grid_n; i++)
			if (g_grid_d[i] == d) { seen = 1; break; }
		if (seen)
			continue;
		g_grid_d[g_grid_n] = d;
		g_grid_c[g_grid_n] = (unsigned char)code;
		g_grid_n++;
	}
	for (i = 1; i < g_grid_n; i++) {
		unsigned d = g_grid_d[i];
		unsigned char c = g_grid_c[i];
		for (j = i - 1; j >= 0 && g_grid_d[j] < d; j--) {
			g_grid_d[j + 1] = g_grid_d[j];
			g_grid_c[j + 1] = g_grid_c[j];
		}
		g_grid_d[j + 1] = d;
		g_grid_c[j + 1] = c;
	}
	g_grid_min = g_grid_d[g_grid_n - 1];
}

/* Greedy largest-fit decomposition of `target` Mac ticks into grid codes.
 * Writes codes to out (up to cap), returns the count or -1 on overflow;
 * *emitted receives the ticks actually covered. */
static long quantize(long target, unsigned char *out, long cap, long *emitted)
{
	long left = target, n = 0;
	int i;
	while (left >= (long)g_grid_min) {
		for (i = 0; i < g_grid_n; i++) {
			if ((long)g_grid_d[i] <= left) {
				if (n >= cap)
					return -1;
				out[n++] = g_grid_c[i];
				left -= (long)g_grid_d[i];
				break;
			}
		}
	}
	if (left > (long)(g_grid_min / 2)) {
		if (n >= cap)
			return -1;
		out[n++] = g_grid_c[g_grid_n - 1];
		left -= (long)g_grid_min;
	}
	*emitted = target - left;
	return n;
}

/* ---- XMI parsing --------------------------------------------------------- */

struct note {
	long tick;
	long dur;
	unsigned char ch, note;
	unsigned char seq;	/* channel first-appearance rank, filled later */
};

static long rd_vlq(const unsigned char *ev, long n, long *p, long *out)
{
	long v = 0;
	unsigned c;
	do {
		if (*p >= n)
			return -1;
		c = ev[*p];
		(*p)++;
		v = (v << 7) | (long)(c & 0x7F);
	} while (c & 0x80);
	*out = v;
	return 0;
}

/* Parse the EVNT chunk into notes[] (cap entries). Returns the note count or
 * XMI2SLB_ERR_*; *tempo receives the last tempo meta seen (default 500000). */
static long parse_xmi(const unsigned char *data, long len, struct note *notes,
		      long cap, unsigned long *tempo)
{
	long i, n, p = 0, t = 0, count = 0;
	const unsigned char *ev;

	*tempo = TEMPO_REF;
	for (i = 0; i + 8 <= len; i++)
		if (data[i] == 'E' && data[i + 1] == 'V' && data[i + 2] == 'N'
		    && data[i + 3] == 'T')
			break;
	if (i + 8 > len)
		return XMI2SLB_ERR_BAD;
	n = ((long)data[i + 4] << 24) | ((long)data[i + 5] << 16)
	  | ((long)data[i + 6] << 8) | (long)data[i + 7];
	ev = data + i + 8;
	if (n > len - (i + 8))
		n = len - (i + 8);		/* the Python slices; a short file just ends */

	while (p < n) {
		unsigned b = ev[p];
		if (b < 0x80) {			/* interval byte (bare, not VLQ) */
			t += (long)b;
			p++;
			continue;
		}
		if (b == 0xFF) {		/* meta */
			unsigned meta;
			long ln2, q;
			if (p + 1 >= n)
				return XMI2SLB_ERR_BAD;
			meta = ev[p + 1];
			q = p + 2;
			if (rd_vlq(ev, n, &q, &ln2) < 0)
				return XMI2SLB_ERR_BAD;
			if (meta == 0x51) {
				if (q + 3 > n)
					return XMI2SLB_ERR_BAD;
				*tempo = ((unsigned long)ev[q] << 16)
				       | ((unsigned long)ev[q + 1] << 8)
				       | (unsigned long)ev[q + 2];
			}
			if (meta == 0x2F)
				break;
			p = q + ln2;
			continue;
		}
		{
			unsigned st = b & 0xF0, ch = b & 0x0F;
			if (st == 0x90) {	/* note on + VLQ duration, no note-offs */
				long q, dur;
				if (p + 3 > n)
					return XMI2SLB_ERR_BAD;
				q = p + 3;
				if (rd_vlq(ev, n, &q, &dur) < 0)
					return XMI2SLB_ERR_BAD;
				if (count >= cap)
					return XMI2SLB_ERR_SPACE;
				notes[count].tick = t;
				notes[count].ch = (unsigned char)ch;
				notes[count].note = ev[p + 1];
				notes[count].dur = dur;
				count++;
				p = q;
			} else if (st == 0x80 || st == 0xA0 || st == 0xB0 || st == 0xE0) {
				p += 3;
			} else if (st == 0xC0 || st == 0xD0) {
				p += 2;
			} else {
				p += 1;
			}
		}
	}
	return count;
}

/* ---- per-voice pattern --------------------------------------------------- */

/* Ordering of the reference's `sorted(events)` over (start, note, dur). */
static int ev_less(const struct note *a, const struct note *b)
{
	if (a->tick != b->tick) return a->tick < b->tick;
	if (a->note != b->note) return a->note < b->note;
	return a->dur < b->dur;
}

static void sort_events(struct note *v, long n)
{
	long i, j;
	for (i = 1; i < n; i++) {		/* insertion sort: n is small per channel */
		struct note x = v[i];
		for (j = i - 1; j >= 0 && ev_less(&x, &v[j]); j--)
			v[j + 1] = v[j];
		v[j + 1] = x;
	}
}

/* monophonize: truncate overlaps, last note wins, drop zero-length. In place;
 * returns the surviving count. */
static long monophonize(struct note *v, long n)
{
	long i, o = 0;
	sort_events(v, n);
	for (i = 0; i < n; i++) {
		if (o > 0 && v[o - 1].tick + v[o - 1].dur > v[i].tick) {
			long d = v[i].tick - v[o - 1].tick;
			v[o - 1].dur = d > 0 ? d : 0;
		}
		v[o++] = v[i];
	}
	/* filter dur > 0 (the reference filters AFTER the whole pass) */
	{
		long w = 0;
		for (i = 0; i < o; i++)
			if (v[i].dur > 0)
				v[w++] = v[i];
		return w;
	}
}

static long put_pair(unsigned char *dst, long cap, long *n, unsigned a, unsigned b)
{
	if (*n + 2 > cap)
		return -1;
	dst[(*n)++] = (unsigned char)a;
	dst[(*n)++] = (unsigned char)b;
	return 0;
}

/* One monophonic channel -> a Mac pattern stream at dst. Events are aimed
 * at their ABSOLUTE ideal Mac tick so quantization error never drifts. */
static long voice_pattern(const struct note *ev, long n, unsigned level,
			  unsigned char *dst, long cap, unsigned char *codes, long ccap)
{
	long out = 0, pos = 0, i, k, nc, emitted;
	static const unsigned char pre[8] = { 0x87, 0x00, 0x83, 0x00, 0x81, 0, 0x82, 0x1A };
	if (cap < 8)
		return -1;
	for (i = 0; i < 8; i++)
		dst[i] = pre[i];
	dst[5] = (unsigned char)level;
	out = 8;
	for (i = 0; i < n; i++) {
		long ideal = ev[i].tick * XMI_TICK;
		if (ideal > pos) {			/* rest-fill the gap */
			nc = quantize(ideal - pos, codes, ccap, &emitted);
			if (nc < 0)
				return -1;
			for (k = 0; k < nc; k++)
				if (put_pair(dst, cap, &out, 128, codes[k]) < 0)
					return -1;
			pos += emitted;
		}
		nc = quantize(ev[i].dur * XMI_TICK, codes, ccap, &emitted);
		if (nc < 0)
			return -1;
		for (k = 0; k < nc; k++) {
			unsigned tie = (k + 1 < nc) ? 0x40 : 0;
			if (put_pair(dst, cap, &out, ev[i].note, codes[k] | tie) < 0)
				return -1;
		}
		pos += emitted;
	}
	if (put_pair(dst, cap, &out, 0xFF, 0xFF) < 0)
		return -1;
	return out;
}

/* ---- song --------------------------------------------------------------- */

static void be16(unsigned char *p, unsigned v) { p[0] = (unsigned char)(v >> 8); p[1] = (unsigned char)v; }

/* period = round(15360 * 500000 / tempo), the reference's round() being
 * half-to-even on an exact quotient. Done as an exact rational so the C
 * cannot disagree with the float by a unit on a .5 boundary. */
static unsigned period_of(unsigned long tempo)
{
	unsigned long long N = (unsigned long long)PERIOD_REF * TEMPO_REF;
	unsigned long long q, r;
	if (tempo == 0)
		tempo = TEMPO_REF;
	q = N / tempo;
	r = N % tempo;
	if (2 * r > tempo || (2 * r == tempo && (q & 1)))
		q++;
	return (unsigned)(q & 0xffffu);
}

long xmi2slb_song(const unsigned char *xmi, long len,
		  unsigned char *dst, long dstcap,
		  unsigned char *scratch, long scratchcap)
{
	struct note *notes, *chan;
	unsigned char *codes, *vbuf;
	long ncap, n, i, nv = 0, out;
	unsigned long tempo;
	long cnt[16], first[16], order[16];
	int pick[MAX_VOICES], npick = 0, c, k;
	long voff[5], vlen[MAX_VOICES], vpos = 0;

	grid_init();

	/* Carve the arena ADAPTIVELY. The first cut gave the whole-file note
	 * array a fixed quarter of the arena, which is 4096 notes at 12 bytes
	 * on a 68000 — and 2730 on a host with 8-byte longs, where the two
	 * Roland-arrangement modules (3591 note-ons, 15.9 KB files) failed with
	 * ERR_SPACE while the 68000 build would have scraped through by luck.
	 * Now: notes take up to HALF the arena; the per-channel copy is sized
	 * from the busiest channel AFTER parsing, placed right behind the notes
	 * actually used; the code buffer and the voice output take the rest. */
	ncap = (scratchcap / 2) / (long)sizeof(struct note);
	if (ncap < 16)
		return XMI2SLB_ERR_SPACE;
	notes = (struct note *)(void *)scratch;

	n = parse_xmi(xmi, len, notes, ncap, &tempo);
	if (n < 0)
		return n;

	/* channels: count, and rank by FIRST APPEARANCE (the reference iterates
	 * a dict, i.e. insertion order, when breaking count ties). Channel 9 is
	 * GM percussion — no pitch — and is dropped. */
	for (c = 0; c < 16; c++) { cnt[c] = 0; first[c] = -1; }
	for (i = 0; i < n; i++) {
		c = notes[i].ch;
		if (c == 9)
			continue;
		if (first[c] < 0)
			first[c] = i;
		cnt[c]++;
	}
	/* candidate list in first-appearance order */
	{
		long m = 0;
		for (i = 0; i < n; i++) {
			c = notes[i].ch;
			if (c != 9 && first[c] == i)
				order[m++] = c;
		}
		/* stable selection: highest count first, ties by appearance */
		for (k = 0; k < MAX_VOICES && k < m; k++) {
			long best = -1, bi = -1;
			for (i = 0; i < m; i++) {
				int used = 0, u;
				for (u = 0; u < npick; u++)
					if (pick[u] == order[i]) used = 1;
				if (used)
					continue;
				if (cnt[order[i]] > best) { best = cnt[order[i]]; bi = i; }
			}
			if (bi < 0)
				break;
			pick[npick++] = (int)order[bi];
		}
	}
	/* keep the source (numeric) ordering of the chosen channels */
	for (k = 1; k < npick; k++) {
		int x = pick[k], j;
		for (j = k - 1; j >= 0 && pick[j] > x; j--)
			pick[j + 1] = pick[j];
		pick[j + 1] = x;
	}

	/* the rest of the arena, behind the notes actually parsed */
	{
		long maxc = 0, used;
		for (c = 0; c < 16; c++)
			if (cnt[c] > maxc) maxc = cnt[c];
		chan = notes + n;
		used = (long)((unsigned char *)(chan + maxc) - scratch);
		if (used + 4096 + 64 > scratchcap)
			return XMI2SLB_ERR_SPACE;
		codes = (unsigned char *)(void *)(chan + maxc);
		vbuf  = codes + 4096;
	}

	/* each voice: copy the channel's events, monophonize, encode */
	for (k = 0; k < npick; k++) {
		long m = 0, l;
		for (i = 0; i < n; i++)
			if (notes[i].ch == pick[k])
				chan[m++] = notes[i];
		m = monophonize(chan, m);
		l = voice_pattern(chan, m, PREAMBLE_LEVELS[k], vbuf + vpos,
				  (scratch + scratchcap) - (vbuf + vpos), codes, 4096);
		if (l < 0)
			return XMI2SLB_ERR_SPACE;
		vlen[k] = l;
		voff[k] = vpos;
		vpos += l;
		nv++;
	}
	for (k = (int)nv; k < 5; k++)
		voff[k] = 0;

	out = 14 + vpos;
	if (out > dstcap)
		return XMI2SLB_ERR_SPACE;
	be16(dst, period_of(tempo));
	dst[2] = 127;
	dst[3] = (unsigned char)nv;
	for (k = 0; k < 5; k++)
		be16(dst + 4 + 2 * k, (unsigned)voff[k]);
	for (i = 0; i < vpos; i++)
		dst[14 + i] = vbuf[i];
	(void)vlen;
	return out;
}

static long empty_song(unsigned char *dst, long cap)
{
	int k;
	if (cap < 16)
		return XMI2SLB_ERR_SPACE;
	be16(dst, (unsigned)PERIOD_REF);
	dst[2] = 127;
	dst[3] = 1;
	for (k = 0; k < 5; k++)
		be16(dst + 4 + 2 * k, 0);
	dst[14] = 0xFF;
	dst[15] = 0xFF;
	return 16;
}

/* ---- bank --------------------------------------------------------------- */

long xmi2slb_bank(const unsigned char *const xmi[3], const long len[3],
		  unsigned char *dst, long dstcap,
		  unsigned char *scratch, long scratchcap)
{
	long hdr = 4 + 4 + 1 + 1 + 2 * (NSLOTS + 1);
	long pos = hdr, off[NSLOTS + 1], total;
	int s;
	if (dstcap < hdr)
		return XMI2SLB_ERR_SPACE;
	for (s = 0; s < NSLOTS; s++) {
		int q = SLOT_MAP[s];
		long l;
		off[s] = pos - hdr;
		if (q != 0 && xmi[q - 1] != 0)
			l = xmi2slb_song(xmi[q - 1], len[q - 1], dst + pos, dstcap - pos,
					 scratch, scratchcap);
		else
			l = empty_song(dst + pos, dstcap - pos);
		if (l < 0)
			return l;
		pos += l;
	}
	off[NSLOTS] = pos - hdr;
	total = pos;
	dst[0] = 'S'; dst[1] = 'L'; dst[2] = 'B'; dst[3] = 'R';
	dst[4] = (unsigned char)(total >> 24); dst[5] = (unsigned char)(total >> 16);
	dst[6] = (unsigned char)(total >> 8);  dst[7] = (unsigned char)total;
	dst[8] = NSLOTS;
	dst[9] = 0;
	for (s = 0; s <= NSLOTS; s++)
		be16(dst + 10 + 2 * s, (unsigned)off[s]);
	return total;
}

/* ---- filename classification -------------------------------------------- */

static int up(int c) { return (c >= 'a' && c <= 'z') ? c - 32 : c; }
static int isdig(int c) { return c >= '0' && c <= '9'; }

int xmi2slb_classify(const char *b, int *drv, int *q)
{
	char u[16];
	int n = 0;
	while (b[n] && n < 15) { u[n] = (char)up((unsigned char)b[n]); n++; }
	u[n] = 0;
	if (n < 8 || u[n - 4] != '.' || u[n - 3] != 'X' || u[n - 2] != 'M' || u[n - 1] != 'I')
		return 0;
	if (n == 9 && u[2] == 'D' && u[3] == 'Q' && isdig(u[4])) {	/* ??DQ<n>.XMI */
		if (u[0] == 'T' && u[1] == 'Y') *drv = XMI2SLB_DRV_TY;
		else if (u[0] == 'P' && u[1] == 'C') *drv = XMI2SLB_DRV_PC;
		else if (u[0] == 'R' && u[1] == 'O') *drv = XMI2SLB_DRV_RO;
		else if (u[0] == 'A' && u[1] == 'D') *drv = XMI2SLB_DRV_AD;
		else return 0;	/* an unknown prefix is not one of the sets (the
				 * reference records it but never selects it) */
		*q = u[4] - '0';
		return 1;
	}
	if (n == 9 && u[0] == 'D' && u[1] == 'Q' && u[2] == 'K' && u[3] == 'Q' && isdig(u[4])) {
		*drv = XMI2SLB_DRV_DQK; *q = u[4] - '0'; return 1;		/* DQKQ<n>.XMI */
	}
	if (n == 8 && u[0] == 'D' && u[1] == 'Q' && u[2] == 'K' && isdig(u[3])) {
		*drv = XMI2SLB_DRV_DQK; *q = u[3] - '0'; return 1;		/* DQK<n>.XMI */
	}
	return 0;
}
